drivers/flash/flash_stm32l5x.c

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright (c) 2021 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_DOMAIN flash_stm32l5
+#define LOG_LEVEL CONFIG_FLASH_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_DOMAIN);
+
+#include <kernel.h>
+#include <device.h>
+#include <string.h>
+#include <drivers/flash.h>
+#include <init.h>
+#include <soc.h>
+#include <stm32l5xx_ll_icache.h>
+#include <stm32_ll_system.h>
+
+#include "flash_stm32.h"
+
+#define STM32L5_SERIES_MAX_FLASH	512
+#define BANK2_OFFSET	(KB(STM32L5_SERIES_MAX_FLASH) / 2)
+
+#define ICACHE_DISABLE_TIMEOUT_VALUE           1U   /* 1ms */
+#define ICACHE_INVALIDATE_TIMEOUT_VALUE        1U   /* 1ms */
+
+static int stm32l5x_icache_disable(void)
+{
+	int status = 0;
+	uint32_t tickstart;
+
+	LOG_DBG("I-cache Disable");
+	/* Clear BSYENDF flag first and then disable the instruction cache that starts
+	 * a cache invalidation procedure
+	 */
+	CLEAR_BIT(ICACHE->FCR, ICACHE_FCR_CBSYENDF);
+
+	CLEAR_BIT(ICACHE->CR, ICACHE_CR_EN);
+
+	/* Get tick */
+	tickstart = HAL_GetTick();
```

Comment:
Instead of using HAL API `HAL_GetTick()`
please use zephyr API: `k_uptime_get_32()`